### PR TITLE
#20946 Add bind address

### LINKED ISSUE #20946: Artery Remoting considerations for ease of use inside Docker
We should make sure it's a) possible and b) easy to run Akka in containers (remember the bind-hostname story etc).

### PR DESCRIPTION
**Pull request by 2m**
Fixes #20946

### CODE REVIEW

**Comment by johanandren** [comment]
This will require changes in a few of the cluster tests as well (which will not run by default so you won't see failures from jenkins)

**Comment by johanandren** [comment]
LGTM although I'm not entierly sold on the ".canonical." part. I like them where they were better.

File: akka-remote/src/main/resources/reference.conf [hunk 1]
```diff
@@ -83,17 +83,48 @@ akka {
...
-      # InetAddress.getLocalHost.getHostAddress is used if empty or
-      # "<getHostAddress>" is specified.
-      # InetAddress.getLocalHost.getHostName is used if
-      # "<getHostName>" is specified.
-      hostname = "<getHostAddress>"
+      canonical {
+        # The default remote server port clients should connect to.
+        # Default is 20200, use 0 if you want a random available port
+        # This port needs to be unique for each actor system on the same machine.
+        port = 20200
```
Comment (patriknw) [inline akka-remote/src/main/resources/reference.conf]: by the way, is this the port we want to use as default?
25520 ?

File: akka-remote/src/main/scala/akka/remote/BoundAddressesExtension.scala [hunk 2]
```diff
@@ -27,6 +28,8 @@ class BoundAddressesExtension(val system: ExtendedActorSystem) extends Extension
    * Returns a mapping from a protocol to a set of bound addresses.
    */
   def boundAddresses: Map[String, Set[Address]] = system.provider
-    .asInstanceOf[RemoteActorRefProvider].transport
-    .asInstanceOf[Remoting].boundAddresses
+    .asInstanceOf[RemoteActorRefProvider].transport match {
+      case artery: ArteryTransport ⇒ Map(("artery" → Set(artery.bindAddress.address)))
```
Comment (patriknw) [inline akka-remote/src/main/scala/akka/remote/BoundAddressesExtension.scala]: `ArteryTransport.ProtocolName`

File: akka-remote/src/main/scala/akka/remote/artery/ArterySettings.scala [hunk 3]
```diff
@@ -22,12 +22,29 @@ private[akka] final class ArterySettings private (config: Config) {
...
-  val Hostname: String = getString("hostname") match {
-    case "" | "<getHostAddress>" ⇒ InetAddress.getLocalHost.getHostAddress
-    case "<getHostName>"         ⇒ InetAddress.getLocalHost.getHostName
-    case other                   ⇒ other
+
+  object Canonical {
+    val config = getConfig("canonical")
+    import config._
+
+    val Port: Int = getInt("port")
```
Comment (patriknw) [inline akka-remote/src/main/scala/akka/remote/artery/ArterySettings.scala]: validation

**Comment by patriknw** [comment]
looking good

File: akka-remote/src/main/resources/reference.conf [hunk 4]
```diff
@@ -83,17 +83,48 @@ akka {
...
       enabled = off
-      port = 20200
 
-      # The hostname or ip clients should connect to.
-      # InetAddress.getLocalHost.getHostAddress is used if empty or
-      # "<getHostAddress>" is specified.
-      # InetAddress.getLocalHost.getHostName is used if
-      # "<getHostName>" is specified.
-      hostname = "<getHostAddress>"
+      canonical {
```
Comment (drewhk) [inline akka-remote/src/main/resources/reference.conf]: hmm, what about swapping:

```
port.canonical = ...
port.bind = ...
hostname.canonical = ...
hostname.bind = ...
```

Comment (drewhk) [inline akka-remote/src/main/resources/reference.conf]: Also, explain what canonical means.

**Comment by drewhk** [comment]
Looks good, my only question is the layout of the config keys.

**Comment by drewhk** [comment]
From a usability perspective probably I would prefer if the following is possible:

If canonical and bind equals:
```
port = ...
hostname  = ...
```

if different:
```
port.canonical = ...
port.bind = ...
hostname.canonical =...
hostname.bind = ...
```

**Comment by 2m** [comment]
That is an interesting config layout. Should be possible with the config library.

File: akka-remote/src/main/resources/reference.conf (hunk 4 above)
Comment (patriknw) [inline akka-remote/src/main/resources/reference.conf]: or just keep it flat?
```
canonical-port
canonical-hostname
bind-port
bind-hostname
```

**Comment by patriknw** [comment]
I think that would be difficult, with default values. Let's not complicate this.

**Comment by drewhk** [comment]
I am not sure if that is difficult, the default can be to set the top level ones and keep the refinements empty. My thinking was to make things straightforward for those that do not need this feature (trying out Akka for example) and still not too hard for those who need it.

**Comment by patriknw** [comment]
I guess it must be tried out, but letting a property sometimes be a section and sometimes a leaf feels like asking for trouble.

**Comment by drewhk** [comment]
Ok, let's keep it simple for now, we can change it later.

File: akka-remote/src/main/resources/reference.conf [hunk 5]
```diff
@@ -93,9 +93,9 @@ akka {
       canonical {
 
         # The default remote server port clients should connect to.
-        # Default is 20200, use 0 if you want a random available port
+        # Default is 25520, use 0 if you want a random available port
         # This port needs to be unique for each actor system on the same machine.
-        port = 20200
+        port = 25520
```
Comment (patriknw) [inline akka-remote/src/main/resources/reference.conf]: well, it was just a crazy suggestion, any opinion @drewhk @johanandren @jrudolph @ktoso 

(fetching paint bucket)

File: akka-remote/src/main/scala/akka/remote/artery/ArterySettings.scala [hunk 6]
```diff
@@ -37,7 +38,7 @@ private[akka] final class ArterySettings private (config: Config) {
 
     val Port: Int = getString("port") match {
       case ""    ⇒ Canonical.Port
-      case other ⇒ getInt("port")
+      case other ⇒ getInt("port").requiring(port ⇒ 0 to 65535 contains port, "bind.port must be 0 through 65535")
```
Comment (patriknw) [inline akka-remote/src/main/scala/akka/remote/artery/ArterySettings.scala]: nice

File: akka-remote/src/main/resources/reference.conf (hunk 5 above)
Comment (johanandren) [inline akka-remote/src/main/resources/reference.conf]: Good to not keep it the default for the "old remoting" I guess. That's my only opinion. Unless we can make it 666 ofc. ;)

**Comment by patriknw** [comment]
@2m looks like a simple test failure, akka vs artery
akka://sys@127.0.1.1:52426 did not equal artery://sys@127.0.1.1:52426

**Comment by patriknw** [comment]
LGTM

**Pull request merged**